### ISSUE TITLE
Support substitutions within custom messages

### ISSUE THREAD

**Issue by onigoetz** [stylelint/stylelint#836]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

I want to set a custom message for `selector-class-pattern` but the default message is a callback with a substituted parameter.
If I set a callback as a parameter, the whole function is transformed to a string

> Which rule, if any, is this issue related to?

Tested on `selector-class-pattern` but almost certainly does it also on any other rule with substitutions

> What CSS is needed to reproduce this issue?

```css
.my_selector {
    margin-top: 3px;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
      "selector-class-pattern": [
        /^(_)?(?:(?:u|t|(?:i|ha)?s)+-[a-z]+[a-zA-Z0-9]*|[A-Z][a-zA-Z0-9]*(__[a-z][a-zA-Z0-9]+)*(--[a-z][a-zA-Z0-9]+)?)$/,
        {
          message: function expected(selectorValue) {
            return "Expected class selector \"." + selectorValue + "\" to match specified pattern, read more here : http://...";
          }
        }
      ],
  }
}
```

> Which version of stylelint are you using?

`4.4.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with the config shown before

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

no

> What did you expect to happen?

I expected the warnings to be formatted like this :

```shell
test.css
1:1     Expected class selector ".my_selector" to match specified pattern, read more here : http://... (selector-class-pattern)
```

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```shell
test.css
1:1     function expected(selectorValue) {
            return "Expected class selector \"." + selectorValue + "\" to match specified pattern, read more here : http://...";
          }

```

**Comment by jeddy3** [comment]
FYI, there's a [discussion here](https://github.com/stylelint/stylelint/issues/400#issuecomment-152832800) about why we started with just simple strings for custom messages.

@davidtheclark Perhaps we should revisit if/how it is possible to support substituted parameter?

@onigoetz With https://github.com/stylelint/stylelint/issues/400 in mind, do you have any thoughts?

**Comment by davidtheclark** [comment]
We could pass all the arguments as an array that get interpolated like so `"Expected {0} to be {1}"`.

But that is also opening a can of worms. Because as people get access to some info in the custom messages they will of course want *more* info. The arguments currently are just based on what we use in the standard message, not on any careful consideration of what *might* be useful in an *arbitrary* message about the warning.

I think that if we really want to make custom messages a powerful feature, we need to do a lot of refactoring.

For each warning we should attach to the warning whatever information might be useful for any arbitrary message about the warning (e.g. which particular selector is in violation, what the original expectation is, etc.). Then we'd provide a means of interpolating that information into the custom messages. 

I'd also consider making the definition of custom messages available in its own place in the config, i.e. you *could* pass it as a secondary option, or you could specify a `customMessages` object that is peer to `rules`. This would enable people to create extendable configs that are just sets of custom messages.

Anyway, this is all to say that I think there's some good work to be done if we want to upgrade support for custom messages. I wonder if anybody is interested in helping make that happen?

**Comment by onigoetz** [comment]
Well, you can't exactly ask people to not want new features ;-)

This can be done on a case by case basis, provide what you have today, and if new informations are available, provide them. I don't think a big rewrite of each rule to provide more informations is a good idea.

As for your ideas, I like the interpolation like `"Expected {0} to be {1}"` almost all translation libraries do it like that, it's a good approach.

But the idea of providing an map of `customMessages` might be a complementary way to add messages (for translations for example) but the message override in the rule should still have priority. It's way cleaner to have the rule, configuration and message at the same place. IMHO

If we go in the direction of interpolation I can try to implement it, tell me if it's a thing you would want and I can try to start.

BTW, that may be a case for an other issue, but I found the default error reporter to be a bit insufficient : no difference between error and warning, and formatting could be improved a bit.
What do you think of a reporter like the one in eslint ?
![capture d ecran 2016-03-03 a 10 10 28](https://cloud.githubusercontent.com/assets/309594/13489597/3b81c212-e128-11e5-9e51-6038cefbe0e3.png)

**Comment by jeddy3** [comment]
Thanks for pulling the reporter issue into a [separate issue](https://github.com/stylelint/stylelint/issues/842) btw :)

**Comment by jeddy3** [comment]
I've pulled the displaying severity into a separate issue.

**Comment by davidtheclark** [comment]
> Well, you can't exactly ask people to not want new features ;-)

I didn't intend to suggest that, sorry if I did.

> This can be done on a case by case basis, provide what you have today, and if new informations are available, provide them. I don't think a big rewrite of each rule to provide more informations is a good idea.

I disagree. I am open to a proof of concept PR demonstrating how this could be done well on a case-by-case basis; but I think that to really make the custom messages feature as powerful as it would need to be to do interpolation right, we need to improve the system behind it. Again, I'm open to a PoC showing that a piecemeal solution would actually work well.

**Comment by onigoetz** [comment]
No it was just a joke, because you said 
> But that is also opening a can of worms. Because as people get access to some info in the custom messages they will of course want more info.

I will experiment in a PR what we discussed

**Comment by davidtheclark** [comment]
If anybody is interested in spearheading this effort and making fully customizable messages available, please pipe up! I think that the implementation is fairly easy and we just need a little organization and some grunt work to get it done.

**Comment by jeddy3** [comment]
@Nirazul Sorry that we haven't had the time [to answer your questions](https://github.com/stylelint/stylelint/pull/865#issuecomment-216537409).  Feel free to continue forward with the idea in the PoC *as you see fit*. Hopefully we'll have time to revisit this once `v7` is done.

**Comment by davidtheclark** [comment]
Closing because it's been a long time since anyone showed enough interest in this to work on it.
